File: android/src/main/kotlin/com/example/acs_flutter_sdk/AcsFlutterSdkPlugin.kt
package com.example.acs_flutter_sdk

import android.content.Context
import com.azure.android.communication.calling.*
import com.azure.android.communication.chat.*
import com.azure.android.communication.common.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.ExecutionException

/**
 * AcsFlutterSdkPlugin
 *
 * Flutter plugin for Microsoft Azure Communication Services.
 * Provides access to calling, chat, and identity management features.
 */
class AcsFlutterSdkPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    // Azure Communication Services clients
    private var tokenCredential: CommunicationTokenCredential? = null
    private var callClient: CallClient? = null
    private var callAgent: CallAgent? = null
    private var call: Call? = null
    private var chatClient: ChatClient? = null
    private var chatThreadClient: ChatThreadClient? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "acs_flutter_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            // Platform info
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            // Identity methods
            "initializeIdentity" -> initializeIdentity(call, result)
            "createUser" -> createUser(result)
            "getToken" -> getToken(call, result)
            "revokeToken" -> revokeToken(call, result)

            // Calling methods
            "initializeCalling" -> initializeCalling(call, result)
            "startCall" -> startCall(call, result)
            "joinCall" -> joinCall(call, result)
            "endCall" -> endCall(result)
            "muteAudio" -> muteAudio(result)
            "unmuteAudio" -> unmuteAudio(result)
            "startVideo" -> startVideo(result)
            "stopVideo" -> stopVideo(result)

            // Chat methods
            "initializeChat" -> initializeChat(call, result)
            "createChatThread" -> createChatThread(call, result)
            "joinChatThread" -> joinChatThread(call, result)
            "sendMessage" -> sendMessage(call, result)
            "getMessages" -> getMessages(call, result)
            "sendTypingNotification" -> sendTypingNotification(result)

            else -> result.notImplemented()
        }
    }

    // ========== Identity Management ==========

    /**
     * Initialize identity client with connection string
     */
    private fun initializeIdentity(call: MethodCall, result: Result) {
        try {
            val connectionString = call.argument<String>("connectionString")
            if (connectionString == null) {
                result.error("INVALID_ARGUMENT", "Connection string is required", null)
                return
            }

            // Note: In production, identity operations should be done server-side
            // This is a simplified implementation for demonstration
            result.success(mapOf("status" to "initialized"))
        } catch (e: Exception) {
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    /**
     * Create a new communication user
     * Note: This should be done server-side in production
     */
    private fun createUser(result: Result) {
        result.error(
            "NOT_IMPLEMENTED",
            "User creation should be done server-side for security. Use your backend API.",
            null
        )
    }

    /**
     * Get access token for a user
     * Note: This should be done server-side in production
     */
    private fun getToken(call: MethodCall, result: Result) {
        result.error(
            "NOT_IMPLEMENTED",
            "Token generation should be done server-side for security. Use your backend API.",
            null
        )
    }

    /**
     * Revoke access token
     */
    private fun revokeToken(call: MethodCall, result: Result) {
        result.error(
            "NOT_IMPLEMENTED",
            "Token revocation should be done server-side. Use your backend API.",
            null
        )
    }

    // ========== Calling ==========

    private fun initializeCalling(call: MethodCall, result: Result) {
        val accessToken = call.argument<String>("accessToken")
        if (accessToken == null) {
            result.error("INVALID_ARGUMENT", "Access token is required", null)
            return
        }

        try {
            // Initialize call client
            callClient = CallClient()

            // Create token credential
            tokenCredential = CommunicationTokenCredential(accessToken)

            // Create call agent
            val callAgentFuture = callClient!!.createCallAgent(
                context,
                tokenCredential!!
            )

            callAgentFuture.whenComplete { agent, error ->
                if (error != null) {
                    result.error("INITIALIZATION_ERROR", error.message, null)
                    return@whenComplete
                }

                callAgent = agent
                result.success(mapOf("status" to "initialized"))
            }
        } catch (e: Exception) {
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun startCall(call: MethodCall, result: Result) {
        val participants = call.argument<List<String>>("participants")
        if (participants == null) {
            result.error("INVALID_ARGUMENT", "Participants list is required", null)
            return
        }

        if (callAgent == null) {
            result.error("NOT_INITIALIZED", "Call agent not initialized. Call initializeCalling first.", null)
            return
        }

        val withVideo = call.argument<Boolean>("withVideo") ?: false

        try {
            // Create call identifiers
            val callees = participants.map { CommunicationUserIdentifier(it) }

            // Start call options
            val startCallOptions = StartCallOptions()

            // Start the call
            this.call = callAgent!!.startCall(context, callees, startCallOptions)

            if (this.call != null) {
                result.success(
                    mapOf(
                        "callId" to this.call!!.id,
                        "state" to callStateToString(this.call!!.state)
                    )
                )
            } else {
                result.error("CALL_START_FAILED", "Failed to start call", null)
            }
        } catch (e: Exception) {
            result.error("CALL_START_FAILED", e.message, null)
        }
    }

    private fun joinCall(call: MethodCall, result: Result) {
        val groupCallId = call.argument<String>("groupCallId")
        if (groupCallId == null) {
            result.error("INVALID_ARGUMENT", "Group call ID is required", null)
            return
        }

        if (callAgent == null) {
            result.error("NOT_INITIALIZED", "Call agent not initialized. Call initializeCalling first.", null)
            return
        }

        val withVideo = call.argument<Boolean>("withVideo") ?: false

        try {
            // Join call options
            val joinCallOptions = JoinCallOptions()

            // Join the call
            val groupCallLocator = GroupCallLocator(java.util.UUID.fromString(groupCallId))
            this.call = callAgent!!.join(context, groupCallLocator, joinCallOptions)

            if (this.call != null) {
                result.success(
                    mapOf(
                        "callId" to this.call!!.id,
                        "state" to callStateToString(this.call!!.state)
                    )
                )
            } else {
                result.error("CALL_JOIN_FAILED", "Failed to join call", null)
            }
        } catch (e: Exception) {
            result.error("CALL_JOIN_FAILED", e.message, null)
        }
    }

    private fun endCall(result: Result) {
        if (call == null) {
            result.error("NO_ACTIVE_CALL", "No active call to end", null)
            return
        }

        try {
            val hangUpFuture = call!!.hangUp(HangUpOptions())
            hangUpFuture.whenComplete { _, error ->
                if (error != null) {
                    result.error("HANGUP_FAILED", error.message, null)
                } else {
                    call = null
                    result.success(null)
                }
            }
        } catch (e: Exception) {
            result.error("HANGUP_FAILED", e.message, null)
        }
    }

    private fun muteAudio(result: Result) {
        if (call == null) {
            result.error("NO_ACTIVE_CALL", "No active call", null)
            return
        }

        try {
            val muteFuture = call!!.muteOutgoingAudio()
            muteFuture.whenComplete { _, error ->
                if (error != null) {
                    result.error("MUTE_FAILED", error.message, null)
                } else {
                    result.success(null)
                }
            }
        } catch (e: Exception) {
            result.error("MUTE_FAILED", e.message, null)
        }
    }

    private fun unmuteAudio(result: Result) {
        if (call == null) {
            result.error("NO_ACTIVE_CALL", "No active call", null)
            return
        }

        try {
            val unmuteFuture = call!!.unmuteOutgoingAudio()
            unmuteFuture.whenComplete { _, error ->
                if (error != null) {
                    result.error("UNMUTE_FAILED", error.message, null)
                } else {
                    result.success(null)
                }
            }
        } catch (e: Exception) {
            result.error("UNMUTE_FAILED", e.message, null)
        }
    }

    private fun startVideo(result: Result) {
        result.error("NOT_IMPLEMENTED", "Video functionality requires camera access and additional setup", null)
    }

    private fun stopVideo(result: Result) {
        result.error("NOT_IMPLEMENTED", "Video functionality requires camera access and additional setup", null)
    }

    // Helper function to convert call state to string
    private fun callStateToString(state: CallState): String {
        return when (state) {
            CallState.NONE -> "none"
            CallState.CONNECTING -> "connecting"
            CallState.RINGING -> "ringing"
            CallState.CONNECTED -> "connected"
            CallState.LOCAL_HOLD -> "onHold"
            CallState.DISCONNECTING -> "disconnecting"
            CallState.DISCONNECTED -> "disconnected"
            CallState.EARLY_MEDIA -> "earlyMedia"
            CallState.REMOTE_HOLD -> "remoteHold"
            else -> "unknown"
        }
    }

    // ========== Chat ==========

    private fun initializeChat(call: MethodCall, result: Result) {
        val accessToken = call.argument<String>("accessToken")
        val endpoint = call.argument<String>("endpoint")

        if (accessToken == null || endpoint == null) {
            result.error("INVALID_ARGUMENT", "Access token and endpoint are required", null)
            return
        }

        try {
            // Create token credential
            tokenCredential = CommunicationTokenCredential(accessToken)

            // Create chat client builder
            val chatClientBuilder = ChatClientBuilder()
                .endpoint(endpoint)
                .credential(tokenCredential!!)

            // Build chat client
            chatClient = chatClientBuilder.buildClient()

            result.success(mapOf("status" to "initialized"))
        } catch (e: Exception) {
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun createChatThread(call: MethodCall, result: Result) {
        val topic = call.argument<String>("topic")
        if (topic == null) {
            result.error("INVALID_ARGUMENT", "Topic is required", null)
            return
        }

        if (chatClient == null) {
            result.error("NOT_INITIALIZED", "Chat client not initialized. Call initializeChat first.", null)
            return
        }

        val participants = call.argument<List<String>>("participants") ?: emptyList()

        try {
            // Create thread request
            val createChatThreadOptions = CreateChatThreadOptions()
                .setTopic(topic)

            // Add participants
            participants.forEach { participantId ->
                val participant = ChatParticipant()
                participant.communicationIdentifier = CommunicationUserIdentifier(participantId)
                createChatThreadOptions.addParticipant(participant)
            }

            // Create thread
            val createChatThreadResult = chatClient!!.createChatThread(createChatThreadOptions)

            if (createChatThreadResult != null) {
                val chatThread = createChatThreadResult.chatThread
                result.success(
                    mapOf(
                        "threadId" to chatThread.id,
                        "topic" to (chatThread.topic ?: "")
                    )
                )
            } else {
                result.error("CREATE_THREAD_FAILED", "Failed to create chat thread", null)
            }
        } catch (e: Exception) {
            result.error("CREATE_THREAD_FAILED", e.message, null)
        }
    }

    private fun joinChatThread(call: MethodCall, result: Result) {
        val threadId = call.argument<String>("threadId")
        if (threadId == null) {
            result.error("INVALID_ARGUMENT", "Thread ID is required", null)
            return
        }

        if (chatClient == null) {
            result.error("NOT_INITIALIZED", "Chat client not initialized. Call initializeChat first.", null)
            return
        }

        try {
            // Get chat thread client
            chatThreadClient = chatClient!!.getChatThreadClient(threadId)
            result.success(mapOf("status" to "joined"))
        } catch (e: Exception) {
            result.error("JOIN_THREAD_FAILED", e.message, null)
        }
    }

    private fun sendMessage(call: MethodCall, result: Result) {
        val threadId = call.argument<String>("threadId")
        val content = call.argument<String>("content")

        if (threadId == null || content == null) {
            result.error("INVALID_ARGUMENT", "Thread ID and content are required", null)
            return
        }

        if (chatClient == null) {
            result.error("NOT_INITIALIZED", "Chat client not initialized", null)
            return
        }

        try {
            val threadClient = chatClient!!.getChatThreadClient(threadId)

            // Send message options
            val sendChatMessageOptions = SendChatMessageOptions()
                .setContent(content)
                .setType(ChatMessageType.TEXT)

            // Send message
            val sendMessageResult = threadClient.sendMessage(sendChatMessageOptions)

            if (sendMessageResult != null) {
                result.success(sendMessageResult.id)
            } else {
                result.error("SEND_MESSAGE_FAILED", "Failed to send message", null)
            }
        } catch (e: Exception) {
            result.error("SEND_MESSAGE_FAILED", e.message, null)
        }
    }

    private fun getMessages(call: MethodCall, result: Result) {
        val threadId = call.argument<String>("threadId")
        if (threadId == null) {
            result.error("INVALID_ARGUMENT", "Thread ID is required", null)
            return
        }

        if (chatClient == null) {
            result.error("NOT_INITIALIZED", "Chat client not initialized", null)
            return
        }

        try {
            val threadClient = chatClient!!.getChatThreadClient(threadId)
            val maxMessages = call.argument<Int>("maxMessages") ?: 20

            // List messages options
            val listMessagesOptions = ListChatMessagesOptions()
                .setMaxPageSize(maxMessages)

            // Get messages
            val messagesPagedIterable = threadClient.listMessages(listMessagesOptions, null)

            val messageList = mutableListOf<Map<String, Any>>()
            messagesPagedIterable.iterableByPage().forEach { page ->
                page.elements.forEach { chatMessage ->
                    messageList.add(
                        mapOf(
                            "id" to chatMessage.id,
                            "content" to (chatMessage.content?.message ?: ""),
                            "senderId" to (chatMessage.senderCommunicationIdentifier?.rawId ?: ""),
                            "sentOn" to (chatMessage.createdOn?.toString() ?: "")
                        )
                    )
                }
            }

            result.success(messageList)
        } catch (e: Exception) {
            result.error("GET_MESSAGES_FAILED", e.message, null)
        }
    }

    private fun sendTypingNotification(result: Result) {
        if (chatThreadClient == null) {
            result.error("NOT_INITIALIZED", "Chat thread client not initialized. Join a thread first.", null)
            return
        }

        try {
            chatThreadClient!!.sendTypingNotification()
            result.success(null)
        } catch (e: Exception) {
            result.error("TYPING_NOTIFICATION_FAILED", e.message, null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        tokenCredential = null
    }
}

File: ios/Classes/AcsFlutterSdkPlugin.swift
import Flutter
import UIKit
import AzureCommunicationCommon
import AzureCommunicationCalling
import AzureCommunicationChat

/// AcsFlutterSdkPlugin
///
/// Flutter plugin for Microsoft Azure Communication Services.
/// Provides access to calling, chat, and identity management features.
public class AcsFlutterSdkPlugin: NSObject, FlutterPlugin {

    // Azure Communication Services clients
    private var tokenCredential: CommunicationTokenCredential?
    private var callClient: CallClient?
    private var callAgent: CallAgent?
    private var call: Call?
    private var chatClient: ChatClient?
    private var chatThreadClient: ChatThreadClient?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "acs_flutter_sdk", binaryMessenger: registrar.messenger())
        let instance = AcsFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // Platform info
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        // Identity methods
        case "initializeIdentity":
            initializeIdentity(call: call, result: result)
        case "createUser":
            createUser(result: result)
        case "getToken":
            getToken(call: call, result: result)
        case "revokeToken":
            revokeToken(call: call, result: result)

        // Calling methods
        case "initializeCalling":
            initializeCalling(call: call, result: result)
        case "startCall":
            startCall(call: call, result: result)
        case "joinCall":
            joinCall(call: call, result: result)
        case "endCall":
            endCall(result: result)
        case "muteAudio":
            muteAudio(result: result)
        case "unmuteAudio":
            unmuteAudio(result: result)
        case "startVideo":
            startVideo(result: result)
        case "stopVideo":
            stopVideo(result: result)

        // Chat methods
        case "initializeChat":
            initializeChat(call: call, result: result)
        case "createChatThread":
            createChatThread(call: call, result: result)
        case "joinChatThread":
            joinChatThread(call: call, result: result)
        case "sendMessage":
            sendMessage(call: call, result: result)
        case "getMessages":
            getMessages(call: call, result: result)
        case "sendTypingNotification":
            sendTypingNotification(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Identity Management

    /// Initialize identity client with connection string
    private func initializeIdentity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let connectionString = args["connectionString"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Connection string is required",
                              details: nil))
            return
        }

        // Note: In production, identity operations should be done server-side
        result(["status": "initialized"])
    }

    /// Create a new communication user
    /// Note: This should be done server-side in production
    private func createUser(result: @escaping FlutterResult) {
        result(FlutterError(code: "NOT_IMPLEMENTED",
                          message: "User creation should be done server-side for security. Use your backend API.",
                          details: nil))
    }

    /// Get access token for a user
    /// Note: This should be done server-side in production
    private func getToken(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(FlutterError(code: "NOT_IMPLEMENTED",
                          message: "Token generation should be done server-side for security. Use your backend API.",
                          details: nil))
    }

    /// Revoke access token
    private func revokeToken(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(FlutterError(code: "NOT_IMPLEMENTED",
                          message: "Token revocation should be done server-side. Use your backend API.",
                          details: nil))
    }

    // MARK: - Calling

    private func initializeCalling(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Access token is required",
                              details: nil))
            return
        }

        do {
            // Initialize call client
            self.callClient = CallClient()

            // Create token credential
            self.tokenCredential = try CommunicationTokenCredential(token: accessToken)

            // Create call agent
            self.callClient?.createCallAgent(userCredential: self.tokenCredential!) { callAgent, error in
                if let error = error {
                    result(FlutterError(code: "INITIALIZATION_ERROR",
                                      message: error.localizedDescription,
                                      details: nil))
                    return
                }

                self.callAgent = callAgent
                result(["status": "initialized"])
            }
        } catch {
            result(FlutterError(code: "INITIALIZATION_ERROR",
                              message: error.localizedDescription,
                              details: nil))
        }
    }

    private func startCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let participants = args["participants"] as? [String] else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Participants list is required",
                              details: nil))
            return
        }

        guard let callAgent = self.callAgent else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Call agent not initialized. Call initializeCalling first.",
                              details: nil))
            return
        }

        let withVideo = args["withVideo"] as? Bool ?? false

        // Create call identifiers
        let callees = participants.map { CommunicationUserIdentifier($0) }

        // Start call options
        let startCallOptions = StartCallOptions()

        // Start the call
        self.call = callAgent.startCall(participants: callees, options: startCallOptions)

        if let call = self.call {
            result([
                "callId": call.id,
                "state": callStateToString(call.state)
            ])
        } else {
            result(FlutterError(code: "CALL_START_FAILED",
                              message: "Failed to start call",
                              details: nil))
        }
    }

    private func joinCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let groupCallId = args["groupCallId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Group call ID is required",
                              details: nil))
            return
        }

        guard let callAgent = self.callAgent else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Call agent not initialized. Call initializeCalling first.",
                              details: nil))
            return
        }

        let withVideo = args["withVideo"] as? Bool ?? false

        // Join call options
        let joinCallOptions = JoinCallOptions()

        // Join the call
        let groupCallLocator = GroupCallLocator(groupId: UUID(uuidString: groupCallId)!)
        self.call = callAgent.join(with: groupCallLocator, joinCallOptions: joinCallOptions)

        if let call = self.call {
            result([
                "callId": call.id,
                "state": callStateToString(call.state)
            ])
        } else {
            result(FlutterError(code: "CALL_JOIN_FAILED",
                              message: "Failed to join call",
                              details: nil))
        }
    }

    private func endCall(result: @escaping FlutterResult) {
        guard let call = self.call else {
            result(FlutterError(code: "NO_ACTIVE_CALL",
                              message: "No active call to end",
                              details: nil))
            return
        }

        call.hangUp(options: HangUpOptions()) { error in
            if let error = error {
                result(FlutterError(code: "HANGUP_FAILED",
                                  message: error.localizedDescription,
                                  details: nil))
            } else {
                self.call = nil
                result(nil)
            }
        }
    }

    private func muteAudio(result: @escaping FlutterResult) {
        guard let call = self.call else {
            result(FlutterError(code: "NO_ACTIVE_CALL",
                              message: "No active call",
                              details: nil))
            return
        }

        call.muteOutgoingAudio() { error in
            if let error = error {
                result(FlutterError(code: "MUTE_FAILED",
                                  message: error.localizedDescription,
                                  details: nil))
            } else {
                result(nil)
            }
        }
    }

    private func unmuteAudio(result: @escaping FlutterResult) {
        guard let call = self.call else {
            result(FlutterError(code: "NO_ACTIVE_CALL",
                              message: "No active call",
                              details: nil))
            return
        }

        call.unmuteOutgoingAudio() { error in
            if let error = error {
                result(FlutterError(code: "UNMUTE_FAILED",
                                  message: error.localizedDescription,
                                  details: nil))
            } else {
                result(nil)
            }
        }
    }

    private func startVideo(result: @escaping FlutterResult) {
        result(FlutterError(code: "NOT_IMPLEMENTED",
                          message: "Video functionality requires camera access and additional setup",
                          details: nil))
    }

    private func stopVideo(result: @escaping FlutterResult) {
        result(FlutterError(code: "NOT_IMPLEMENTED",
                          message: "Video functionality requires camera access and additional setup",
                          details: nil))
    }

    // Helper function to convert call state to string
    private func callStateToString(_ state: CallState) -> String {
        switch state {
        case .none:
            return "none"
        case .connecting:
            return "connecting"
        case .ringing:
            return "ringing"
        case .connected:
            return "connected"
        case .localHold:
            return "onHold"
        case .disconnecting:
            return "disconnecting"
        case .disconnected:
            return "disconnected"
        case .earlyMedia:
            return "earlyMedia"
        case .remoteHold:
            return "remoteHold"
        @unknown default:
            return "unknown"
        }
    }

    // MARK: - Chat

    private func initializeChat(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let endpoint = args["endpoint"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Access token and endpoint are required",
                              details: nil))
            return
        }

        do {
            // Create token credential
            self.tokenCredential = try CommunicationTokenCredential(token: accessToken)

            // Create chat client configuration
            let chatClientOptions = AzureCommunicationChatClientOptions()

            // Initialize chat client
            self.chatClient = try ChatClient(
                endpoint: endpoint,
                credential: self.tokenCredential!,
                withOptions: chatClientOptions
            )

            result(["status": "initialized"])
        } catch {
            result(FlutterError(code: "INITIALIZATION_ERROR",
                              message: error.localizedDescription,
                              details: nil))
        }
    }

    private func createChatThread(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let topic = args["topic"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Topic is required",
                              details: nil))
            return
        }

        guard let chatClient = self.chatClient else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Chat client not initialized. Call initializeChat first.",
                              details: nil))
            return
        }

        let participants = args["participants"] as? [String] ?? []

        // Create thread request
        let request = CreateChatThreadRequest(
            topic: topic,
            participants: participants.map { participantId in
                ChatParticipant(
                    id: CommunicationUserIdentifier(participantId),
                    displayName: nil,
                    shareHistoryTime: nil
                )
            }
        )

        chatClient.create(thread: request) { response, error in
            if let error = error {
                result(FlutterError(code: "CREATE_THREAD_FAILED",
                                  message: error.localizedDescription,
                                  details: nil))
                return
            }

            if let chatThread = response?.chatThread {
                result([
                    "threadId": chatThread.id,
                    "topic": chatThread.topic ?? ""
                ])
            } else {
                result(FlutterError(code: "CREATE_THREAD_FAILED",
                                  message: "Failed to create chat thread",
                                  details: nil))
            }
        }
    }

    private func joinChatThread(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let threadId = args["threadId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Thread ID is required",
                              details: nil))
            return
        }

        guard let chatClient = self.chatClient else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Chat client not initialized. Call initializeChat first.",
                              details: nil))
            return
        }

        do {
            // Get chat thread client
            self.chatThreadClient = try chatClient.createClient(forThread: threadId)
            result(["status": "joined"])
        } catch {
            result(FlutterError(code: "JOIN_THREAD_FAILED",
                              message: error.localizedDescription,
                              details: nil))
        }
    }

    private func sendMessage(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let threadId = args["threadId"] as? String,
              let content = args["content"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Thread ID and content are required",
                              details: nil))
            return
        }

        guard let chatClient = self.chatClient else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Chat client not initialized",
                              details: nil))
            return
        }

        do {
            let threadClient = try chatClient.createClient(forThread: threadId)

            let request = SendChatMessageRequest(
                content: content,
                senderDisplayName: nil,
                type: .text,
                metadata: nil
            )

            threadClient.send(message: request) { response, error in
                if let error = error {
                    result(FlutterError(code: "SEND_MESSAGE_FAILED",
                                      message: error.localizedDescription,
                                      details: nil))
                    return
                }

                if let messageId = response?.id {
                    result(messageId)
                } else {
                    result(FlutterError(code: "SEND_MESSAGE_FAILED",
                                      message: "Failed to send message",
                                      details: nil))
                }
            }
        } catch {
            result(FlutterError(code: "SEND_MESSAGE_FAILED",
                              message: error.localizedDescription,
                              details: nil))
        }
    }

    private func getMessages(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let threadId = args["threadId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                              message: "Thread ID is required",
                              details: nil))
            return
        }

        guard let chatClient = self.chatClient else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Chat client not initialized",
                              details: nil))
            return
        }

        do {
            let threadClient = try chatClient.createClient(forThread: threadId)
            let maxMessages = args["maxMessages"] as? Int ?? 20

            let options = ListChatMessagesOptions(maxPageSize: maxMessages)

            threadClient.listMessages(withOptions: options) { response, error in
                if let error = error {
                    result(FlutterError(code: "GET_MESSAGES_FAILED",
                                      message: error.localizedDescription,
                                      details: nil))
                    return
                }

                if let messages = response?.items {
                    let messageList = messages.compactMap { message -> [String: Any]? in
                        guard let chatMessage = message as? ChatMessage else { return nil }

                        return [
                            "id": chatMessage.id,
                            "content": chatMessage.content?.message ?? "",
                            "senderId": chatMessage.sender?.identifier ?? "",
                            "sentOn": ISO8601DateFormatter().string(from: chatMessage.createdOn ?? Date())
                        ]
                    }
                    result(messageList)
                } else {
                    result([])
                }
            }
        } catch {
            result(FlutterError(code: "GET_MESSAGES_FAILED",
                              message: error.localizedDescription,
                              details: nil))
        }
    }

    private func sendTypingNotification(result: @escaping FlutterResult) {
        guard let chatThreadClient = self.chatThreadClient else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Chat thread client not initialized. Join a thread first.",
                              details: nil))
            return
        }

        chatThreadClient.sendTypingNotification() { error in
            if let error = error {
                result(FlutterError(code: "TYPING_NOTIFICATION_FAILED",
                                  message: error.localizedDescription,
                                  details: nil))
            } else {
                result(nil)
            }
        }
    }
}
